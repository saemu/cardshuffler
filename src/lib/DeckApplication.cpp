/*
 * Copyright 2020, Samuel Brand
 */

#include "DeckApplication.hpp"

#include "lib/Card.hpp"

#include <algorithm>
#include <map>
#include <spdlog/fmt/ostr.h>
#include <spdlog/spdlog.h>
#include <string>

namespace {
std::string readInput(std::istream& input, spdlog::logger& output) {
    std::string quit;
    output.info("new shuffle [y/n]: ");
    if (std::getline(input, quit)) {
        quit.erase(std::remove_if(quit.begin(), quit.end(), ::isspace), quit.end());
    }
    return quit;
}

bool validateInput(const std::string& input, const std::map<std::string, bool>& validInputs) {
    auto const answer = validInputs.find(input);
    if (answer == std::cend(validInputs)) {
        throw std::invalid_argument("your answer is wrong, exiting");
    }
    return answer->second;
}

bool readAnswer(std::istream& input, spdlog::logger& output) {
    // clang-format off
    static const std::map<std::string, bool> validInputs{{"y",   true,},
                                                         {"yes", true,},
                                                         {"n",   false},
                                                         {"no",  false},};
    // clang-format on
    std::string answer = readInput(input, output);
    return validateInput(answer, validInputs);
}

}  // namespace

void DeckApplication::run(std::istream& input, spdlog::logger& output) {
    bool running{true};
    Deck deck;
    do {
        deck.shuffle();
        do {
            const Card card = deck.drawNextCard();
            output.info("Card {}", card);
        } while (deck.hasNext());
        running = readAnswer(input, output);
    } while (running);
}
