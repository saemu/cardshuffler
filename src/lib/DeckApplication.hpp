/*
 * Copyright 2020, Samuel Brand
 */

#pragma once

#include "lib/Deck.hpp"

namespace spdlog {
class logger;
}

class DeckApplication {
public:
    static void run(std::istream& input, spdlog::logger& output);
};
