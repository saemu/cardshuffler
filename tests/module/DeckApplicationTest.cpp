#include <gtest/gtest.h>
#include <spdlog/sinks/ostream_sink.h>
#include <spdlog/spdlog.h>

#include <ApprovalTests.hpp>

#include "DeckApplication.hpp"

using namespace ApprovalTests;

struct DeckApplicationTest : testing::Test {
    void SetUp() override {
        input.clear();
        output.clear();
        sink = std::make_shared<spdlog::sinks::ostream_sink_mt>(output);
        sink->set_pattern("%v");
        logger = std::make_shared<spdlog::logger>("testlogger", sink);
    }

    void TearDown() override {
        spdlog::drop_all();
    }

    std::shared_ptr<spdlog::logger> logger = nullptr;
    std::shared_ptr<spdlog::sinks::sink> sink = nullptr;
    std::stringstream input;
    std::stringstream output;
};

TEST_F(DeckApplicationTest, ValidInput) {
    input << R"__(n
)__";
    DeckApplication::run(0U, input, *logger);
    Approvals::verify(output.str());
}

TEST_F(DeckApplicationTest, InvalidInput) {
    input << R"__(foo
)__";

    EXPECT_THROW(DeckApplication::run(0U, input, *logger), std::invalid_argument);
    Approvals::verify(output.str());
}
