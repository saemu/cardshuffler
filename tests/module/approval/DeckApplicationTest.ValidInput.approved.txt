Card Seven of Spades
Card Six of Spades
Card Eight of Hearts
Card Six of Clubs
Card Nine of Clubs
Card Ten of Hearts
Card Ace of Spades
Card Seven of Hearts
Card Ace of Clubs
Card Queen of Spades
Card Six of Diamonds
Card King of Diamonds
Card Jack of Clubs
Card Queen of Clubs
Card Seven of Diamonds
Card Ten of Diamonds
Card Eight of Clubs
Card King of Hearts
Card King of Clubs
Card Jack of Spades
Card Nine of Spades
Card Seven of Clubs
Card Ten of Clubs
Card Ace of Diamonds
Card Eight of Diamonds
Card Six of Hearts
Card Jack of Diamonds
Card Ten of Spades
Card Nine of Diamonds
Card Eight of Spades
Card Queen of Diamonds
Card King of Spades
Card Nine of Hearts
Card Queen of Hearts
Card Ace of Hearts
Card Jack of Hearts
new shuffle [y/n]: 

